### ISSUE TITLE
TransformerEncoderLayer returns very different results on float64

### ISSUE THREAD

**Issue by twoertwein** [pytorch/pytorch#146128]
### 🐛 Describe the bug

`TransformerEncoderLayer` should return very similar results when run with float32 and float64, but they are very different:

```py
import torch

torch.manual_seed(0)
model = torch.nn.TransformerEncoderLayer(32, 1)
x = torch.rand(10, 32)

y = model(x)

y64 = model.to(dtype=torch.float64)(x.to(dtype=torch.float64))

# should get very similar results with float64
print((y - y64).abs().max().item())  # but get a giant difference: 2.3634471505652055
```

(I was using float64 to debug some numerical discrepencies.)

### Versions

Collecting environment information...
PyTorch version: 2.6.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.1.1 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.4)
CMake version: version 3.31.5
Libc version: N/A

Python version: 3.11.10 | packaged by conda-forge | (main, Sep 10 2024, 10:57:35) [Clang 17.0.6 ] (64-bit runtime)
Python platform: macOS-15.1.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] mypy==1.11.2
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] onnx==1.17.0
[pip3] onnx2torch==1.5.15
[pip3] onnxconverter-common==1.14.0
[pip3] onnxmltools==1.12.0
[pip3] onnxruntime==1.19.2
[pip3] onnxscript==0.1.0.dev20241214
[pip3] optree==0.13.0
[pip3] skl2onnx==1.18.0
[pip3] tf2onnx==1.16.1
[pip3] torch==2.6.0
[pip3] torch-onnx==0.1.25
[pip3] torchaudio==2.6.0
[pip3] torchvision==0.20.1
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] onnx2torch                1.5.15                   pypi_0    pypi
[conda] optree                    0.13.0                   pypi_0    pypi
[conda] torch                     2.6.0                    pypi_0    pypi
[conda] torch-onnx                0.1.25                   pypi_0    pypi
[conda] torchaudio                2.6.0                    pypi_0    pypi
[conda] torchvision               0.20.1                   pypi_0    pypi

**Comment by colesbury** [comment]
TransformerEncoderLayer has dropout by default and dropout is random. You will get different results if you call `model(x)` multiple times, regardless of the data type.
